runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java

```diff
@@ -688,53 +712,28 @@ private void checkClusterId(@Nonnull Layout layout) {
                         // we discard it.
-                        if (l.getEpoch() < router.getEpoch()) {
+                        if (latestLayout != null && latestLayout.getEpoch() > l.getEpoch()) {
                             log.warn("fetchLayout: Received a layout with epoch {} from server "
-                                            + "{}:{} smaller than router epoch {}, discarded.",
-                                    l.getEpoch(), router.getHost(),
-                                    router.getPort(), router.getEpoch());
+                                            + "{}:{} smaller than latestLayout epoch {}, "
+                                            + "discarded.",
+                                    l.getEpoch(), router.getHost(), router.getPort(),
+                                    latestLayout.getEpoch());
                             continue;
                         }
 
                         checkClusterId(l);
 
                         l.setRuntime(this);
-                        // this.layout should only be assigned to the new layout future
-                        // once it has been completely constructed and initialized.
-                        // For example, assigning this.layout = l
-                        // before setting the layout's runtime can result in other threads
-                        // trying to access a layout with  a null runtime.
-                        // FIXME Synchronization START
-                        // We are updating multiple variables and we need the update to be
-                        // synchronized across all variables.
-                        // Since the variable layoutServers is used only locally within the class
-                        // it is acceptable (at least the code on 10/13/2016 does not have issues)
-                        // but setEpoch of routers needs to be synchronized as those variables are
-                        // not local.
-                        for (String server : l.getAllServers()) {
-                            try {
-                                getRouter(server).setEpoch(l.getEpoch());
-                            } catch (NetworkException ne) {
-                                // We have already received the layout and there is no need to keep
-                                // client waiting.
-                                // NOTE: This is true assuming this happens only at router creation.
-                                // If not we also have to take care of setting the latest epoch on
-                                // Client Router.
-                                if (!l.getUnresponsiveServers().contains(server)) {
-                                    log.warn("fetchLayout: Error getting router : {}", server, ne);
-                                }
-                            }
-                        }
                         layoutServers = l.getLayoutServers();
+                        latestLayout = l;
```

Comment:
As we discussed you can try to get rid of LayoutServers variable as well. Then there is only one state variable the object.

Reply:
Removed.